addons/payment/models/payment_acquirer.py

```diff
@@ -279,6 +279,20 @@ def button_immediate_install(self):
 
     #=== BUSINESS METHODS ===#
 
+    def _verify_matching_companies(self, record, message=None):
+        """
+        Check that the company in which the acquirer was defined/installed matches
+        the company of the record that is the object of the payment.
+        In the case of a multi-company environment avoids partners paying for documents that
+        do not belong to their company.
+        """
+        if record and record.company_id:
+            if message is None:
+                message = "You must log in with a partner of the company that owns this document."
+
+            if record.company_id != self.company_id:
+                raise UserError(message)
```

Comment:
I'd move this method to `payment`'s portal controllers class. Here, you made it a method on `payment.acquirer` but it has several flaws:
- You make a check on the acquirer's company but that company ultimately comes from the URL or the document.
- You need to get the list of compatible acquirers *before* checking if the company configuration even makes any sense.
- The method should be multi (`for acquirer in self`) but it wouldn't give any additional information since, again, the acquirers' company comes from higher in the flow.

Reply:
Move it to PaymentPortal, now before calling `_get_available_acquirers` we fail if there is a mismatch.